### ISSUE TITLE
[JSON Output]: allow to add more json data

### ISSUE THREAD

**Issue by roderik** [jmcdo29/ogma#297]
## Describe the solution you'd like

In our application, we have a lot of context which is relevant for our logs. 

e.g.

- on which cloud provider did this service execute
- on which unique identifier did this service work
- ...

In general, correlation ids and useful metadata you can explorer in a log viewer.

Currently, there is only 1 field that could be used, the requestID, but we need more and do not want to replace the request correlation id.

[In Winston, they catalog all these as "meta" ](https://github.com/winstonjs/winston#streams-objectmode-and-info-objects)

Now, for output, in json mode, it would output `{...meta, ...currentfields}` while in text mode you either leave them out or print them in a block at the end of thee line.

By adding them to the json output, you can add columns and filter on these fields, see below for an example in datadog (with only the current fields available)

![image](https://user-images.githubusercontent.com/16780/97010046-33bec880-1545-11eb-9939-4e16aa254885.png)

## Teachability, Documentation, Adoption, Migration Strategy

There is no breaking change, a single optional extra param to 'log' would be all that is required.

## What is the motivation / use case for changing the behavior?

see above

**Comment by jmcdo29** [comment]
Thanks for the conversation on GitHub, to recap it here the following changes are probably going to be implemented:

* log methods' signatures will change from `(message: string, context?: string, application?: string, requestId?: string)` to `(message: string, metaInfo: {context?: string, application?: string, correlationId?: string, [key: string]: unknown })` to allow for adding in extra metadata
* in JSON mode, the extra metadata will be added as top level keys so that log aggregators can better display logs with column titles
* text mode will ignore new metadata but will keep the `context`, `application`, and `correlationId`
* `requestId` will be renamed to `correlationId`

Did I miss anything? This will be a breaking change.

**Comment by roderik** [comment]
you could in json mode even keep everything under a "meta" key, that would be just fine in Kibana/Datadog etc

You could add a "debug/verbose" flag to the settings so it prints 2 lines, the second one with the meta data JSON stringified

**Comment by jmcdo29** [comment]
Added to v0.4.0. This will most likely be the last pre v1 version. Thanks for the idea

**Comment by roderik** [comment]
As an FYI for people upgrading, this release is breaking for some usage

```
    this.ogma.debug(`${message}`, component);
```

Now fails with the following typescript error

```
src/services/logger.ts(43,59): error TS2345: Argument of type 'string' is not assignable to parameter of type 'OgmaPrintOptions | undefined'.
src/services/logger.ts(47,59): error TS2345: Argument of type 'string' is not assignable to parameter of type 'OgmaPrintOptions | undefined'.
src/services/logger.ts(58,58): error TS2345: Argument of type 'string' is not assignable to parameter of type 'OgmaPrintOptions | undefined'.
src/services/logger.ts(66,59): error TS2345: Argument of type 'string' is not assignable to parameter of type 'OgmaPrintOptions | undefined'.
src/services/logger.ts(70,58): error TS2345: Argument of type 'string' is not assignable to parameter of type 'OgmaPrintOptions | undefined'.
src/services/apu.ts(290,56): error TS2345: Argument of type 'string' is not assignable to parameter of type '"" | "topic" | "direct" | "headers" | "fanout" | "match"'.
src/server.ts(35,17): error TS2345: Argument of type 'OgmaService' is not assignable to parameter of type 'LoggerService'.
  Types of property 'log' are incompatible.
    Type '(message: any, meta?: OgmaServiceMeta | undefined) => void' is not assignable to type '(message: any, context?: string | undefined) => any'.
      Types of parameters 'meta' and 'context' are incompatible.
        Type 'string | undefined' is not assignable to type 'OgmaServiceMeta | undefined'.
          Type 'string' is not assignable to type 'OgmaServiceMeta | undefined'.
```

Fixed by changing it to

```
    this.ogma.debug(`${message}`, {
      context: component,
    });
```

**Comment by jmcdo29** [comment]
Yep. That was part of the original breaking change of `0.4.0`, but now should be fixed on `0.4.1`
